### PR TITLE
Improve reprs for all objects in API

### PR DESCRIPTION
**Pull request by ardila**
Adds (or changes) all reprs so that

eval(str(test_object)) == test_object

I didn't feel sure about changing what equality meant for Model yet, but everything else I made an equality check that is as extensive as possible.

I added lots of tests for this and tested using

poetry run pytest. Some tests fail due to non-uniqueness issue with model ids in the database, but that happens in master too.

I tried to make things that feel like they are just data containers into dataclasses.

There was some pylint and typing stuff I fixed along the way as well.

Once we have #51 merged, that will make this PR smaller. If you want it split up further let me know.

### CODE REVIEW

**Comment by ardila** [comment]
We should merge some of the other changes first (especially the poetry one) then I can clean things up a bit since I made this change on top of those (by accident)

File: nucleus/__init__.py [hunk 1]
```diff
@@ -961,6 +975,42 @@ def delete_slice(self, slice_id: str) -> dict:
         )
         return response
 
+    def append_to_slice(
```
Comment (ardila) [inline nucleus/__init__.py:978]: I moved this here to stop pylint complaining about directly accessing the client's private method

File: conftest.py [hunk 2]
```diff
@@ -6,6 +6,7 @@
 # https://github.com/gevent/gevent/issues/1016#issuecomment-328530533
 # https://github.com/spyoungtech/grequests/issues/8
 import grequests
+
```
Comment (ardila) [inline conftest.py:9]: These are just changes due to black. If we want to remove them from this PR let me know

File: nucleus/model_run.py [hunk 3]
```diff
@@ -125,19 +134,31 @@ def loc(self, dataset_item_id: str):
...
+                Type[BoxPrediction],
+                Type[PolygonPrediction],
+                Type[SegmentationPrediction],
+            ],
+        ] = {
+            BOX_TYPE: BoxPrediction,
+            POLYGON_TYPE: PolygonPrediction,
+            SEGMENTATION_TYPE: SegmentationPrediction,
+        }
+        for type_key in annotation_payload:
```
Comment (ardila) [inline nucleus/model_run.py:158]: I added this so we stop getting mypy errors about this.

File: nucleus/prediction.py [hunk 4]
```diff
@@ -91,9 +93,6 @@ def from_json(cls, payload: dict):
             metadata=payload.get(METADATA_KEY, {}),
         )
 
-    def __str__(self):
```
Comment (ardila) [inline nucleus/prediction.py:94]: These are no longer needed because of inheritance from the base class

File: nucleus/slice.py [hunk 5]
```diff
@@ -47,18 +51,9 @@ def append(
...
-
-        payload: Dict[str, Any] = {}
-        if dataset_item_ids:
-            payload[DATASET_ITEM_IDS_KEY] = dataset_item_ids
-        if reference_ids:
-            payload[REFERENCE_IDS_KEY] = reference_ids
-
-        response = self._client._make_request(
-            payload, f"slice/{self.slice_id}/append"
+        response = self._client.append_to_slice(
```
Comment (ardila) [inline nucleus/slice.py:54]: Moved the logic into the client to avoid direct access.

**Comment by ardila** [comment]
Ok this should be ready for review!

File: nucleus/annotation.py [hunk 6]
```diff
@@ -33,16 +41,11 @@ def from_json(cls, payload: dict):
             return SegmentationAnnotation.from_json(payload)
 
 
+@dataclass
```
Comment (sasha-scale) [inline nucleus/annotation.py:44]: oh this is cool, never seen dataclass before.

Great for brevity!

File: nucleus/annotation.py [hunk 7]
```diff
@@ -62,35 +65,25 @@ def to_payload(self) -> dict:
...
+    mask_url: str
+    annotations: List[Segment]
+    reference_id: Optional[str] = None
+    item_id: Optional[str] = None
+    annotation_id: Optional[str] = None
+
+    def __post_init__(self):
+        if not self.mask_url:
             raise Exception("You must specify a mask_url.")
-        if bool(reference_id) == bool(item_id):
```
Comment (sasha-scale) [inline nucleus/annotation.py:77]: Is this deletion purposeful? I think UX hypothesis behind this check is that it's much better user experience to know your SegmentationAnnotation is malformed before sending the API request.

File: nucleus/prediction.py [hunk 8]
```diff
@@ -1,3 +1,5 @@
+from __future__ import annotations
```
Comment (sasha-scale) [inline nucleus/prediction.py:1]: This import seems unused to me - can you elaborate on what it's needed for?

**Review by sasha-scale — APPROVED** [review]
LGTM, this is a big improvement for debugging!

File: nucleus/prediction.py (hunk 8 above)
Comment (ardila) [reply nucleus/prediction.py:1]: This allows for the type annotations in this file to work properly.

Comment (sasha-scale) [reply nucleus/prediction.py:1]: why would it be needed here but not other files (like `nucleus/annotation.py`)

**Pull request merged**